quickwit/quickwit-indexing/src/source/ingest/mod.rs

```diff
@@ -0,0 +1,766 @@
+    assigned_shards: HashMap<ShardId, AssignedShard>,
+    fetch_stream: MultiFetchStream,
+    publish_lock: PublishLock,
+    publish_token: PublishToken,
+}
+
+impl fmt::Debug for IngestSource {
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.debug_struct("IngestSource").finish()
+    }
+}
+
+impl IngestSource {
+    pub async fn try_new(
+        runtime_args: Arc<SourceRuntimeArgs>,
+        _checkpoint: SourceCheckpoint,
+    ) -> anyhow::Result<Self> {
+        let node_id: NodeId = runtime_args.node_id().into();
+        let client_id = ClientId::new(
+            node_id.clone(),
+            runtime_args.index_uid().clone(),
+            runtime_args.source_id().to_string(),
+            runtime_args.pipeline_ord(),
+        );
+        let metastore = runtime_args.metastore.clone();
+        let ingester_pool = runtime_args.ingester_pool.clone();
+        let assigned_shards = HashMap::new();
+        let fetch_stream =
+            MultiFetchStream::new(node_id, client_id.client_id(), ingester_pool.clone());
+        let publish_lock = PublishLock::default();
+        let publish_token = format!("{}:{}", client_id.client_id(), Ulid::new());
+
+        Ok(Self {
+            client_id,
+            metastore,
+            ingester_pool,
+            assigned_shards,
+            fetch_stream,
+            publish_lock,
+            publish_token,
+        })
+    }
+
+    fn process_fetch_response(
+        &mut self,
+        batch_builder: &mut BatchBuilder,
+        fetch_response: FetchResponseV2,
+    ) -> anyhow::Result<()> {
+        let assigned_shard = self
+            .assigned_shards
+            .get_mut(&fetch_response.shard_id)
+            .expect("shard should be assigned");
+        let partition_id = assigned_shard.partition_id.clone();
+
+        for doc in fetch_response.docs() {
+            batch_builder.push(doc);
+        }
+        let from_position_exclusive = assigned_shard.current_position_inclusive.clone();
+        let to_position_inclusive = fetch_response
+            .to_position_inclusive()
+            .map(Position::from)
+            .unwrap_or(Position::Beginning);
+        batch_builder
+            .checkpoint_delta
+            .record_partition_delta(
+                partition_id,
+                from_position_exclusive,
+                to_position_inclusive.clone(),
+            )
+            .context("failed to record partition delta")?;
+        assigned_shard.current_position_inclusive = to_position_inclusive;
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl Source for IngestSource {
+    async fn emit_batches(
+        &mut self,
+        doc_processor_mailbox: &Mailbox<DocProcessor>,
+        ctx: &SourceContext,
+    ) -> Result<Duration, ActorExitStatus> {
+        let now = Instant::now();
+        let mut batch_builder = BatchBuilder::default();
+        let deadline = time::sleep(EMIT_BATCHES_TIMEOUT);
+        tokio::pin!(deadline);
```

Comment:
tokio::time::time_out does not work?

Reply:
It does (it's `timeout_at`). All the sources have this defect. I'll fix it when I work on #3829.